[REVIEW] Using new ucx-py API & implementing CUML_ANY_SOURCE semantics on irecv

cjnolet opened the pull request.

teju85 commented on cpp/src_prims/common/cuml_comms_int.hpp:
```diff
@@ -43,6 +43,8 @@ class cumlCommunicator {
   enum datatype_t { CHAR, UINT8, INT, UINT, INT64, UINT64, FLOAT, DOUBLE };
   enum op_t { SUM, PROD, MIN, MAX };
 
+  static const int ANY_RANK = -1;
```
How about exposing this along our cumlcomms interface so that users don't have to redeclare tags like `UCP_ANY_RANK` as done above?

cjnolet replied:
I defined a separate constant only to keep `ucp_helper` decoupled from the comms implementation, since I've seen some interest in using it on its own. I agree, though, creating a separate constant does make this awkward and brittle. I'll think about this one a bit. Definitely agree the ANY_RANK constant should be moved to the interface.

cjnolet commented:
@teju85 I put a `CUML_ANY_SOURCE` constant on the comms interface and both the mpi & std comms are using it now.

cjnolet commented:
@teju85, you'll be happy to know that this PR also builds the std comms by default from the top level build, rather than having to build it separately! I've added the MPI comms to the top level as well, though I've made it optional for now until we discuss making MPI conda packaging a dependency. I've added the options `BUILD_CUML_MPI_COMMS` and `BUILD_CUML_STD_COMMS`

cjnolet commented:
@madsbk, I've integrated your ucx-py changes here. Do you mind giving this a quick review?

cjnolet commented:
@teju85 any chance I could get another review or approval on this one? The updates for the new ucx-py API are needed for us to be able to develop MNMG algos for 0.11.

teju85 commented:
@cjnolet give me until tomorrow to review this one.

teju85 commented on cpp/README.md:
```diff
@@ -34,6 +34,8 @@ Current cmake offers the following configuration options:
 | BUILD_CUML_MG_TESTS | [ON, OFF]  | ON  |  Enable/disable building cuML algorithm test executable `ml_mg_test`. |
 | BUILD_PRIMS_TESTS | [ON, OFF]  | ON  | Enable/disable building cuML algorithm test executable `prims_test`.  |
+| BUILD_CUML_STD_COMMS | [ON, OFF] | ON | Enable/disable building cuML NCCL+UCX communicator for running multi-node multi-GPU algorithms. |
+| BUILD_CUML_MPI_COMMS | [ON, OFF] | OFF | Enable/disable building cuML MPI+NCCP communicator for running multi-node multi-GPU C++ tests. |
```
```suggestion
| BUILD_CUML_MPI_COMMS | [ON, OFF] | OFF | Enable/disable building cuML MPI+NCCL communicator for running multi-node multi-GPU C++ tests. |
```

teju85 commented on cpp/comms/mpi/src/cuML_comms_mpi_impl.cpp:
```diff
@@ -268,6 +271,9 @@ void cumlMPICommunicator_impl::irecv(void *buf, int size, int source, int tag, r
     }
+
+
+
```
minor knitpick: too many newlines

cjnolet commented:
Thanks again @teju85 for reviewing this! I've addressed your comments!
